src/device_manager.rs

```diff
@@ -0,0 +1,365 @@
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
+
+                        return Err(Error::DeviceOverlap);
+                    }
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    if self
+                        .mmio_bus
+                        .insert(IoRange::new_mmio_range(base, size), device.clone())
+                        .is_some()
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
+
+                        return Err(Error::DeviceOverlap);
+                    }
+                }
+                _ => continue,
+            }
+        }
+        Ok(())
+    }
+
+    /// Unregister a device from `IoManager`, e.g. users specified removing.
+    /// VMM pre-fetches the resources e.g. dev.get_assigned_resources()
+    /// VMM is responsible for freeing the resources.
+    ///
+    /// # Arguements
+    ///
+    /// * `resources`: resources that this device owns, might include
+    ///                port I/O and memory-mapped I/O ranges, irq number, etc.
+    pub fn unregister_device_io(&mut self, resources: &[Resource]) -> Result<()> {
+        for res in resources.iter() {
+            match *res {
+                Resource::PioAddressRange { base, size } => {
+                    self.pio_bus.remove(&IoRange::new_pio_range(base, size));
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    self.mmio_bus.remove(&IoRange::new_mmio_range(base, size));
+                }
+                _ => continue,
+            }
+        }
+        Ok(())
+    }
+
+    fn first_before(
+        &self,
+        addr: IoAddress,
+        map: BTreeMap<IoRange, Arc<dyn DeviceIo>>,
+    ) -> Option<(IoRange, Arc<dyn DeviceIo>)> {
+        for (range, dev) in map.iter().rev() {
+            if range.base <= addr {
+                return Some((*range, dev.clone()));
+            }
+        }
+        None
+    }
+
+    // Return the Device mapped the address.
+    fn get_device(
+        &self,
+        addr: IoAddress,
+        map: BTreeMap<IoRange, Arc<dyn DeviceIo>>,
+    ) -> Option<Arc<dyn DeviceIo>> {
+        if let Some((range, dev)) = self.first_before(addr, map) {
+            if (addr.raw_value() - range.base.raw_value()) < range.size.raw_value() {
+                return Some(dev);
+            }
+        }
+        None
+    }
+
+    /// A helper function handling PIO read command during VM exit.
+    /// The virtual device itself provides mutable ability and thead-safe protection.
+    ///
+    /// Return error if failed to get the device.
+    pub fn pio_read(&self, addr: u16, data: &mut [u8]) -> Result<()> {
+        if let Some(device) = self.get_device(IoAddress::Pio(addr), self.pio_bus.clone()) {
```

Comment:
The IoManager::pio_bus() is an instance of BtreeMap, the clone() implies a deep copy. It's really really too heavy here.
Please refer to Clone for BtreeMap:
https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#138-214

Reply:
Solved and removed.